cupy/_core/core.pyx

```diff
@@ -185,14 +201,10 @@ cdef class _ndarray_base:
         cdef tuple s = internal.get_size(shape)
         del shape
 
+        # this would raise if order is not recognized
         cdef int order_char = (
             b'C' if order is None else internal._normalize_order(order))
 
-        # `strides` is prioritized over `order`, but invalid `order` should be
-        # checked even if `strides` is given.
-        if order_char != b'C' and order_char != b'F':
-            raise ValueError('order not understood. order=%s' % order)
```

Comment:
I think this is dead code that we'd never reach, see the comment at line 204 that I added above.

Reply:
@kmaehashi I think I missed the second half of the story (sorry it's been a while since I worked on this feature...). `TypeError` is raised in `__array__`:
https://github.com/cupy/cupy/blob/678add80cdd82e89f2b751f1722035c53997ad03/cupy/_core/core.pyx#L1489-L1493, which can be easily reproduced by adding a dummy `def __array__(self): raise TypeError` to your snippet above. We do not allow implicitly converting CuPy arrays to NumPy (object) arrays.